unison-cli/src/Unison/Codebase/Editor/HandleInput/FormatFile.hs

```diff
@@ -0,0 +1,221 @@
+
+    -- Typechecking ALWAYS adds a type-signature, but we don't want to add ones that didn't
+    -- already exist in the source file.
+    removeGeneratedTypeAnnotations ::
+      UnisonFile Symbol a -> Symbol -> (Term.Term Symbol a) -> (Term.Term Symbol a)
+    removeGeneratedTypeAnnotations uf v = \case
+      Term.Ann' tm _annotation | not (hasUserTypeSignature uf v) -> tm
+      x -> x
+
+    -- This is a bit of a hack.
+    -- The file parser uses a different unique ID for unique types on every parse,
+    -- that id changes hashes, and constructors are ordered by that hash.
+    -- This means that pretty-printing isn't deterministic and constructors will re-order
+    -- themselves on every save :|
+    --
+    -- It's difficult and a bad idea to change the parser to use a deterministic unique ID,
+    -- so instead we just re-sort the constructors by their source-file annotation AFTER
+    -- parsing. This is fine for pretty-printing, but don't use this for anything other than
+    -- formatting since the Decls it produces aren't technically valid.
+    mkUnisonFilesDeterministic :: Maybe (UnisonFile Symbol Ann.Ann) -> Maybe (TypecheckedUnisonFile Symbol Ann.Ann) -> (Maybe (UnisonFile Symbol Ann.Ann), Maybe (TypecheckedUnisonFile Symbol Ann.Ann))
+    mkUnisonFilesDeterministic mayUnisonFile mayTypecheckedFile =
+      let sortedUF =
+            mayUnisonFile
+              & _Just . #dataDeclarationsId . traversed . _2 %~ sortConstructors
+              & _Just . #effectDeclarationsId . traversed . _2 . Decl.asDataDecl_ %~ sortConstructors
+          sortedTF =
+            mayTypecheckedFile
+              & _Just . #dataDeclarationsId' . traversed . _2 %~ sortConstructors
+              & _Just . #effectDeclarationsId' . traversed . _2 . Decl.asDataDecl_ %~ sortConstructors
+       in (sortedUF, sortedTF)
+
+    -- ppedForFileHelper
+    sortConstructors :: Decl.DataDeclaration v Ann.Ann -> Decl.DataDeclaration v Ann.Ann
+    sortConstructors dd =
+      -- Sort by their Ann so we keep the order they were in the original file.
+      dd & Decl.constructors_ %~ sortOn @Ann.Ann (view _1)
+
+annToRange :: Ann.Ann -> Maybe Range
+annToRange = \case
+  Ann.Intrinsic -> Nothing
+  Ann.External -> Nothing
+  Ann.GeneratedFrom a -> annToRange a
+  Ann.Ann start end -> Just $ Range start end
+
+rangeToInterval :: Range -> Interval.Interval Pos.Pos
+rangeToInterval (Range start end) =
+  Interval.ClosedInterval start end
+
+annToInterval :: Ann.Ann -> Maybe (Interval.Interval Pos.Pos)
+annToInterval ann = annToRange ann <&> rangeToInterval
+
+-- | Returns 'True' if the given symbol is a term with a user provided type signature in the
+-- parsed file, false otherwise.
+hasUserTypeSignature :: Eq v => UnisonFile v a -> v -> Bool
+hasUserTypeSignature parsedFile sym =
+  UF.terms parsedFile
+    & any (\(v, _, trm) -> v == sym && isJust (Term.getTypeAnnotation trm))
+
+-- | Apply a list of updates to a text, returning the updated text.
```

Comment:
Could you clarify the haddock about what the function does / what applying an update means? I guess replacing a certain range indexed _[start, end)_ with a `Text`. Might be clearer to swap the pair order if it doesn't make a mess elsewhere.

Reply:
Alternatively, based on the comment on line 30, you considered a newtype `TextEdit`? I think that might be clearer too, but I'm guessing it was a pain to use?